### PR TITLE
Shopping cart

### PR DESCRIPTION
**Pull request by BohdanZorii**

### CODE REVIEW

File: src/main/java/mate/zorii/bookstore/controller/BookController.java [hunk 1]
```diff
@@ -41,7 +41,7 @@ public Page<BookDto> findAll(
     @GetMapping("/{id}")
     @PreAuthorize("hasRole('USER')")
     @Operation(summary = "Get a book by ID", description = "Returns the book with the specified ID")
-    public BookDto findById(@PathVariable @PositiveOrZero Long id) {
+    public BookDto findById(@PathVariable @Positive Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/controller/BookController.java:44]: add `@Validated` under the controller

File: src/main/java/mate/zorii/bookstore/controller/CategoryController.java [hunk 2]
```diff
@@ -42,7 +42,7 @@ public Page<CategoryDto> getAll(Pageable pageable) {
     @PreAuthorize("hasRole('USER')")
     @Operation(summary = "Get category by ID",
             description = "Retrieves a specific category by its ID.")
-    public CategoryDto getCategoryById(@PathVariable @PositiveOrZero Long id) {
+    public CategoryDto getCategoryById(@PathVariable @Positive Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/controller/CategoryController.java:45]: same

File: src/main/java/mate/zorii/bookstore/dto/shoppingcart/CartItemDto.java [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+
+public record CartItemDto(
+        @NotNull(message = "Book ID cannot be null")
+        @Positive(message = "Book ID must be positive")
+        Long bookId,
+        @NotNull(message = "Quantity cannot be null")
+        @Min(value = 1, message = "Quantity must be at least 1")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/dto/shoppingcart/CartItemDto.java:12]:
```suggestion
        @Positive(value = 1, message = "Quantity must be at least 1")
```

File: src/main/java/mate/zorii/bookstore/dto/shoppingcart/CartItemUpdateDto.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package mate.zorii.bookstore.dto.shoppingcart;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+
+public record CartItemUpdateDto(
+        @NotNull(message = "Quantity cannot be null")
+        @Min(value = 1, message = "Quantity must be at least 1")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/dto/shoppingcart/CartItemUpdateDto.java:8]:
```suggestion
        @Positive(value = 1, message = "Quantity must be at least 1")
```

File: src/main/java/mate/zorii/bookstore/model/CartItem.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+@Entity
+@Table(name = "cart_items")
+@Getter
+@Setter
+public class CartItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/CartItem.java:22]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/zorii/bookstore/model/CartItem.java [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    @JoinColumn(name = "shopping_cart_id", referencedColumnName = "id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/CartItem.java:25]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/mate/zorii/bookstore/model/ShoppingCart.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import jakarta.persistence.OneToMany;
+import jakarta.persistence.OneToOne;
+import jakarta.persistence.Table;
+import java.util.HashSet;
+import java.util.Set;
+import lombok.Getter;
+import lombok.Setter;
+
+@Entity
+@Table(name = "shoppingCarts")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/ShoppingCart.java:19]:
```suggestion
@Table(name = "shopping_carts")
```

File: src/main/java/mate/zorii/bookstore/model/ShoppingCart.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+@Getter
+@Setter
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @OneToOne(fetch = FetchType.EAGER)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/ShoppingCart.java:26]:
```suggestion
    @MapsId
    @OneToOne(fetch = FetchType.LAZY)
```

you can use MapsId here

File: src/main/java/mate/zorii/bookstore/repository/CartItemRepository.java [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+
+import java.util.Optional;
+import mate.zorii.bookstore.model.CartItem;
+import mate.zorii.bookstore.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+    Optional<CartItem> findCartItemByBook_IdAndShoppingCart(Long bookId, ShoppingCart shoppingCart);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/repository/CartItemRepository.java:11]: It's better to use
```suggestion
    Optional<CartItem> findByIdAndShoppingCartId(Long cartItemId, Long shoppingCartId);
```

File: src/main/java/mate/zorii/bookstore/repository/ShoppingCartRepository.java [hunk 10]
```diff
@@ -0,0 +1,11 @@
+package mate.zorii.bookstore.repository;
+
+import java.util.Optional;
+import mate.zorii.bookstore.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    Optional<ShoppingCart> findShoppingCartByUser_Id(Long userId);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/repository/ShoppingCartRepository.java:10]:
```suggestion
    @EntityGraph(attributePaths = {"cartItems", "cartItems.book"})
    ShoppingCart findByUserId(Long userId);
```

File: src/main/java/mate/zorii/bookstore/service/ShoppingCartService.java [hunk 11]
```diff
@@ -0,0 +1,15 @@
+package mate.zorii.bookstore.service;
+
+import mate.zorii.bookstore.dto.shoppingcart.CartItemDto;
+import mate.zorii.bookstore.dto.shoppingcart.CartItemUpdateDto;
+import mate.zorii.bookstore.dto.shoppingcart.ShoppingCartResponseDto;
+
+public interface ShoppingCartService {
+    ShoppingCartResponseDto findByUserId(Long user);
+
+    CartItemDto addCartItem(CartItemDto cartItemDto, Long authenticatedUser);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/ShoppingCartService.java:10]:
```suggestion
    ShoppingCartResponseDto addCartItem(CartItemDto cartItemDto, Long authenticatedUser);
```

File: src/main/java/mate/zorii/bookstore/service/ShoppingCartService.java [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+
+    CartItemDto updateCartItem(Long cartItemId, CartItemUpdateDto cartItemUpdateDto);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/ShoppingCartService.java:12]:
```suggestion
    ShoppingCartResponseDto updateCartItem(Long cartItemId, CartItemUpdateDto cartItemUpdateDto);
```

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final CartItemRepository cartItemRepository;
+    private final ShoppingCartMapper mapper;
+
+    @Override
+    public ShoppingCartResponseDto findByUserId(Long userId) {
+        return mapper.toShoppingCartDto(getCartByUserId(userId));
+    }
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:29]: Add `@Transactional` where is needed

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    public CartItemDto updateCartItem(Long cartItemId, CartItemUpdateDto cartItemUpdateDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(() ->
+                new EntityNotFoundException("No cart item by id " + cartItemId));
+        mapper.updateCartItemFromDto(cartItem, cartItemUpdateDto);
+        return mapper.toCartItemDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public void deleteCartItem(Long id) {
+        System.out.println("was here");
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:55]:
```suggestion
```

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        cartItemRepository.deleteById(id);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:56]:
```suggestion
        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(itemId, shoppingCart.getId())........
        cartItemRepository.deleteById(id);
```

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+    @Override
+    public CartItemDto updateCartItem(Long cartItemId, CartItemUpdateDto cartItemUpdateDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(() ->
+                new EntityNotFoundException("No cart item by id " + cartItemId));
+        mapper.updateCartItemFromDto(cartItem, cartItemUpdateDto);
+        return mapper.toCartItemDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public void deleteCartItem(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:54]:
```suggestion
    public void deleteCartItem(Long id, Long userId) {
```

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+        }
+        return mapper.toCartItemDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public CartItemDto updateCartItem(Long cartItemId, CartItemUpdateDto cartItemUpdateDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(() ->
+                new EntityNotFoundException("No cart item by id " + cartItemId));
+        mapper.updateCartItemFromDto(cartItem, cartItemUpdateDto);
+        return mapper.toCartItemDto(cartItemRepository.save(cartItem));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:50]:
```suggestion
        ShoppingCart cart = shoppingCartRepository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, cart.getId())
                .map(item -> {
                    item.setQuantity(itemDto.quantity());
                    return item;
                }).orElseThrow(() -> new EntityNotFoundException(
                        String.format("No cart item with id: %d for user: %d", cartItemId, userId)
                ));
        cartItemRepository.save(cartItem);
        return cartMapper.toDto(cart);
```

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    private final ShoppingCartMapper mapper;
+
+    @Override
+    public ShoppingCartResponseDto findByUserId(Long userId) {
+        return mapper.toShoppingCartDto(getCartByUserId(userId));
+    }
+
+    @Override
+    public CartItemDto addCartItem(CartItemDto cartItemDto, Long userId) {
+        ShoppingCart shoppingCart = getCartByUserId(userId);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:31]:
```
Book book = bookRepository.findById
ShoppingCart cart = repository.findByUserId(userId);
cart.getCartItems().stream() .... 
```
if book exists - increase quantity
if no - add to cart
```
repository.save(cart);
return cartMapper.toDto(cart);
```

File: src/main/java/mate/zorii/bookstore/service/impl/UserServiceImpl.java [hunk 19]
```diff
@@ -23,7 +26,12 @@ public UserResponseDto register(UserRegistrationRequestDto bookDto)
             throw new RegistrationException("Email is already registered");
         }
         User user = userMapper.toModel(bookDto);
-        userRepository.save(user);
+        User savedUser = userRepository.save(user);
+
+        ShoppingCart shoppingCart = new ShoppingCart();
+        shoppingCart.setUser(savedUser);
+        shoppingCartRepository.save(shoppingCart);
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/UserServiceImpl.java:34]:
```suggestion
        userRepository.save(user);
        shoppingCartService.createShoppingCart(user);
```

Where in ShoppingCartServiceImpl

```
@Override
    public void createShoppingCart(User user) {
        ShoppingCart shoppingCart = new ShoppingCart();
        shoppingCart.setUser(user);
        repository.save(shoppingCart);
    }
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution

File: src/main/java/mate/zorii/bookstore/repository/CartItemRepository.java (hunk 9 above)
Comment (BohdanZorii) [reply src/main/java/mate/zorii/bookstore/repository/CartItemRepository.java:11]: I don't get cartItemId from client when creating cartItem. Moreover, I don't see how this checks, if cartItem with the concrete book is already present in the shoppingCart

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+        CartItem cartItem = shoppingCart.getCartItems().stream()
+                .filter(item -> item.getBook().getId().equals(cartItemDto.bookId()))
+                .findFirst()
+                .map(item -> {
+                    item.setQuantity(item.getQuantity() + cartItemDto.quantity());
+                    return item;
+                }).orElse(mapper.toModel(cartItemDto, shoppingCart));
+
+        cartItemRepository.save(cartItem);
+        return findByUserId(userId);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:49]: Could we avoid calling DB ?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
one comment

File: src/main/java/mate/zorii/bookstore/dto/shoppingcart/CartItemDto.java [hunk 21]
```diff
@@ -0,0 +1,14 @@
+package mate.zorii.bookstore.dto.shoppingcart;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+
+public record CartItemDto(
+        @NotNull(message = "Book ID cannot be null")
+        @Positive(message = "Book ID must be positive")
+        Long bookId,
+        @NotNull(message = "Quantity cannot be null")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/dto/shoppingcart/CartItemDto.java:10]:
```suggestion
```
Didn't mention before, primitive type can't be null

File: src/main/java/mate/zorii/bookstore/dto/shoppingcart/CartItemUpdateDto.java [hunk 22]
```diff
@@ -0,0 +1,10 @@
+package mate.zorii.bookstore.dto.shoppingcart;
+
+import jakarta.validation.constraints.NotNull;
+import jakarta.validation.constraints.Positive;
+
+public record CartItemUpdateDto(
+        @NotNull(message = "Quantity cannot be null")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/dto/shoppingcart/CartItemUpdateDto.java:7]:
```suggestion
```

File: src/main/java/mate/zorii/bookstore/model/ShoppingCart.java [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+@Setter
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @MapsId
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "user_id", referencedColumnName = "id", nullable = false)
+    private User user;
+    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/model/ShoppingCart.java:31]:
```suggestion
    @OneToMany(mappedBy = "shoppingCart", cascade = CascadeType.ALL, orphanRemoval = true)
```

Lazy by default

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 24]
```diff
@@ -0,0 +1,84 @@
...
+                }).orElseThrow(() -> new EntityNotFoundException(
+                        String.format("No cart item with id: %d for user: %d",
+                                cartItemId, userId)));
+        cartItemRepository.save(cartItem);
+        return mapper.toShoppingCartDto(shoppingCart);
+    }
+
+    @Override
+    public void deleteCartItem(Long cartItemId, Long userId) {
+        cartItemRepository.findByIdAndShoppingCart_Id(cartItemId, userId)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:71]: You are trying to find cart item by id and shopCart id, but provide cart item id and user id
```suggestion
        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(userId);
        CartItem cartItem = cartItemRepository.findByIdAndShoppingCartId(cartItemId, shoppingCart.getId())
```

File: src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+                    item.setQuantity(cartItemUpdateDto.quantity());
+                    return item;
+                }).orElseThrow(() -> new EntityNotFoundException(
+                        String.format("No cart item with id: %d for user: %d",
+                                cartItemId, userId)));
+        cartItemRepository.save(cartItem);
+        return mapper.toShoppingCartDto(shoppingCart);
+    }
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/ShoppingCartServiceImpl.java:69]:
```suggestion
    @Override
    @Transactional
```

File: src/main/java/mate/zorii/bookstore/service/impl/UserServiceImpl.java [hunk 26]
```diff
@@ -6,15 +6,19 @@
...
 
 @Service
 @RequiredArgsConstructor
 public class UserServiceImpl implements UserService {
     private final UserRepository userRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
     private final UserMapper userMapper;
+    private final ShoppingCartService shoppingCartService;
 
     @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/zorii/bookstore/service/impl/UserServiceImpl.java:23]:
```suggestion
    @Override
    @Transactional
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Well done. Left some comments

File: src/main/java/mate/zorii/bookstore/model/ShoppingCart.java (hunk 23 above)
Comment (BohdanZorii) [reply src/main/java/mate/zorii/bookstore/model/ShoppingCart.java:31]: it was said that it's better to set it explicitly

Comment (Elena-Bruyako) [reply src/main/java/mate/zorii/bookstore/model/ShoppingCart.java]: Not required. Some teams prefer to not use default values due to leave code readable

**Review by Elena-Bruyako — APPROVED** [review]
And we have done with this part eventually )) Congrats

**Pull request merged**